### PR TITLE
vconsole: search for usable source console

### PR DESCRIPTION
**Pull request by msoltyspl**
When vconsole-setup is called without arguments, search for a usable
console instead of using /dev/tty0.

/dev/tty0 - pointing to the current active console - it not necessarily
usable and in such case vconsole-setup would exit with failure.

https://github.com/systemd/systemd/issues/5367

### CODE REVIEW

File: src/vconsole/vconsole-setup.c [hunk 1]
```diff
@@ -328,48 +320,79 @@ static void setup_remaining_vcs(int fd, bool utf8) {
         }
 }
 
+static const char *find_source_vc(const char *vc, int *src_fd, int *src_idx) {
+        static char ttyname[sizeof("/dev/tty") - 1 + DECIMAL_STR_MAX(int)];
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: humm, static variables like this I don't really like I must say. It's fine to use static variables if the data placed there is effectively constant, but doing this in a function that places different data there depending on the parameters is something I don#t like at all I must say.

Sure, this is a private function here, that is not exposed to anything else, but still, I think we really shouldn't even do this.

There's nothing wrong with allocating a string for this on the heap.

I'd much prefer if we'd change this to something like this:

```c
static int find_source_vc(char **ret_path, int *ret_idx);
```

File: src/vconsole/vconsole-setup.c [hunk 2]
```diff
@@ -328,48 +320,79 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+
+        if (vc) {
+                _cleanup_close_ int fd = -1;
+
+                fd = open_terminal(vc, O_RDWR|O_CLOEXEC);
+                if (fd < 0) {
+                        log_error_errno(fd, "Failed to open %s: %m", vc);
+                        return NULL;
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: hmm, i think we should propagate the error as error really and not as NULL. i.e. if the function returns an int then a positive return value would be an fd, and the negative returns would be a negative errno error like we do everywhere else (also see CODING_STYLE about that)

File: src/vconsole/vconsole-setup.c [hunk 3]
```diff
@@ -55,7 +55,7 @@ static bool is_vconsole(int fd) {
 }
 
 static bool is_allocated(unsigned int idx) {
-        char vcname[strlen("/dev/vcs") + DECIMAL_STR_MAX(int)];
+        char vcname[sizeof("/dev/vcs") + DECIMAL_STR_MAX(int) - 2];
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: we don#t actually mind if you call strlen() like this here. In theory it's a function call, and hence this would become a VLA. However, on gcc and llvm (and probably all smart compilers) strlen() on a literal string is optimized away, and constant. They grok that strlen() is what it is...

But using sizeof() is fine too

File: src/vconsole/vconsole-setup.c [hunk 4]
```diff
@@ -328,48 +320,88 @@ static void setup_remaining_vcs(int fd, bool utf8) {
         }
 }
 
+static int find_source_vc(char **ret_path, int *ret_idx, const char *src_vc) {
+        int ret_fd;
+
+        if (src_vc) {
+                _cleanup_close_ int fd = -1;
+
+                if (!(*ret_path = malloc(strlen(src_vc) + 1))) {
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: hmm, please do not clobber return parameters on failure (see CODING_STYLE). i.e. a function should only overwrite return parameters on success, but on failure leave them the way they are.

also, we generally prefer using type-safe `new(char, strlen(src_vc) + 1)` instead of malloc() when allocating place for strings.

Finally, we really try to make things readable, and hence much prefer if each line does one thing. Specifically, I think this should be two lines: one line doing the allocation, and the next line checking if it worked. (also see CODING_STYLE)

File: src/vconsole/vconsole-setup.c [hunk 5]
```diff
@@ -328,48 +320,88 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                        log_oom();
+                        return -1;
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: please don't make up errors such as -1. Please only return negative errno style errors, i.e. "return -ENOMEM" or so. Note that log_oom() returns -ENOMEM anywa,y hence you can simplify the above then to:

```
return log_oom();
```

which is a very typical idiom in the systemd sources

File: src/vconsole/vconsole-setup.c [hunk 6]
```diff
@@ -328,48 +320,88 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                }
+                fd = open_terminal(src_vc, O_RDWR|O_CLOEXEC);
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: we probably want O_NOCTTY here, too, just to be sure?

File: src/vconsole/vconsole-setup.c [hunk 7]
```diff
@@ -328,48 +320,88 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                if (fd < 0) {
+                        log_error_errno(fd, "Failed to open %s: %m", src_vc);
+                        return -errno;
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: open_terminal() returns the error as return value, and you shouldn't use errno here. Also log_error_errno() returns the error passed to it as negative value, hence you can shorten this:

```c
if (fd < 0)
        return log_error_errno(fd, "....");
```

and the right thing will happen

File: src/vconsole/vconsole-setup.c [hunk 8]
```diff
@@ -328,48 +320,88 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                }
+
+                if (!is_vconsole(fd)) {
+                        log_error("Device %s is not a virtual console.", src_vc);
+                        return -1;
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: please don't make up erros. Maybe return -ENOTTY or so.

File: src/vconsole/vconsole-setup.c [hunk 9]
```diff
@@ -328,48 +320,88 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                }
+
+                if (!is_allocated_byfd(fd)) {
+                        log_error("Virtual console %s is not allocated.", src_vc);
+                        return -1;
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: as above...

File: src/vconsole/vconsole-setup.c [hunk 10]
```diff
@@ -328,48 +320,88 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                }
+
+                if (!is_settable(fd)) {
+                        log_error("Virtual console %s is not in K_XLATE or K_UNICODE.", src_vc);
+                        return -1;
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: and here...

File: src/vconsole/vconsole-setup.c [hunk 11]
```diff
@@ -328,48 +320,88 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                }
+
+                strcpy(*ret_path, src_vc);
+                *ret_idx = -1;
+                ret_fd = fd;
+                fd = -1;
+                return ret_fd;
+        }
+
+        if (!(*ret_path = malloc(sizeof("/dev/tty") + DECIMAL_STR_MAX(int) - 2))) {
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: as above...

File: src/vconsole/vconsole-setup.c [hunk 12]
```diff
@@ -328,48 +320,88 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                log_oom();
+                return -1;
+        }
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: as above

File: src/vconsole/vconsole-setup.c [hunk 13]
```diff
@@ -328,48 +320,88 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                        continue;
+
+                *ret_idx = i;
+                ret_fd = fd;
+                fd = -1;
+                return ret_fd;
+        }
+
+        log_error("Unable to find a candidate for the source console.");
+        return -1;
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: similar here, please use a proper error, something appropriate. Maybe -ENXIO or so

Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c]: I will return most recent error from the loop iteration now. That should be most useful.

I'll outline or changes in the original thread after the commit.

File: src/vconsole/vconsole-setup.c (hunk 6 above)
Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c]: Good point. Will update this in other places of the code as well.

File: src/vconsole/vconsole-setup.c [hunk 14]
```diff
@@ -82,7 +84,8 @@ static bool is_settable(int fd) {
          *
          * http://lists.freedesktop.org/archives/systemd-devel/2013-February/008573.html
          */
-        return r == 0 && IN_SET(curr_mode, K_XLATE, K_UNICODE);
+        return r < 0 ? -errno :
+               IN_SET(curr_mode, K_XLATE, K_UNICODE) ? 0 : (errno = EBUSY, -EBUSY);
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c]: No, please don't do that. That's not easy to read — just use normal `if`s.

Don't set errno, we never do that.

File: src/vconsole/vconsole-setup.c [hunk 15]
```diff
@@ -280,22 +275,22 @@ static void setup_remaining_vcs(int fd, bool utf8) {
         if (cfo.op != KD_FONT_OP_SET)
                 log_warning("Fonts will not be copied to remaining consoles");
 
-        for (i = 1; i <= 63; i++) {
-                char ttyname[strlen("/dev/tty") + DECIMAL_STR_MAX(int)];
+        for (unsigned i = 1; i <= 63; i++) {
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c]: Unfortunately Lennart does not like C99, please declare the variable on top of the block.

File: src/vconsole/vconsole-setup.c [hunk 16]
```diff
@@ -305,36 +300,109 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+
+                /* all checks passed, return this one as a source console */
+                *ret_idx = i;
+                *ret_path = path;
+                ret_fd = fd;
+                fd = -1;
+                return ret_fd;
+        }
+
+        errno = loop_errno;
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c]: Here too, please don't set errno. Just return it as a negative error code.

(Whoever came up with the idea of errno, we should all pitch in for an early retirement fund for them, as long at they promise to never touch the computer or design anything again.)

File: src/vconsole/vconsole-setup.c [hunk 17]
```diff
@@ -305,36 +300,109 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                        log_warning_errno(errno, "PIO_UNIMAP failed, unimaps might be incorrect for tty%u: %m", i);
                         continue;
                 }
 
                 log_debug("Font and unimap successfully copied to %s", ttyname);
         }
 }
 
+static int find_source_vc(char **ret_path, unsigned *ret_idx) {
+        char *path;
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c]: `_cleanup_free_ *path = NULL;`

File: src/vconsole/vconsole-setup.c [hunk 18]
```diff
@@ -305,36 +300,109 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                        continue;
+                }
+                if (is_settable(fd) < 0) {
+                        loop_errno = errno;
+                        continue;
+                }
+
+                /* all checks passed, return this one as a source console */
+                *ret_idx = i;
+                *ret_path = path;
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c:366]: `path = NULL;`

File: src/vconsole/vconsole-setup.c [hunk 19]
```diff
@@ -305,36 +300,109 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                ret_fd = fd;
+                fd = -1;
+                return ret_fd;
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c]: `return 0`?

File: src/vconsole/vconsole-setup.c [hunk 20]
```diff
@@ -305,36 +300,109 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+        }
+
+        errno = loop_errno;
+        free(path);
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c]: Not necessary with `_cleanup_`.

File: src/vconsole/vconsole-setup.c [hunk 21]
```diff
@@ -305,36 +300,109 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+        return log_error_errno(errno, "No usable source console found, last error for /dev/tty%u was: %m", i - 1);
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c]: We generally prefer to return the first error, not the last, and don't include the information that there was more than one error. Here too, I expect the first error to be at least as good as the last one.

File: src/vconsole/vconsole-setup.c [hunk 22]
```diff
@@ -305,36 +300,109 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+        if (fd < 0)
+                return log_error_errno(fd, "Failed to open %s: %m", src_vc);
+
+        if (is_vconsole(fd) < 0)
+                return log_error_errno(errno, "Device %s is not a virtual console: %m", src_vc);
+
+        if (is_allocated_byfd(fd) < 0)
+                return log_error_errno(errno, "Virtual console %s is not allocated: %m", src_vc);
+
+        if (is_settable(fd) < 0)
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c]: Hm, shouldn't those conditions all be `<= 0`?

File: src/vconsole/vconsole-setup.c [hunk 23]
```diff
@@ -305,36 +300,109 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                return log_error_errno(errno, "Virtual console %s is not in K_XLATE or K_UNICODE: %m", src_vc);
+
+        path = new(char, strlen(src_vc) + 1);
+        if (path == NULL)
+                return log_oom();
+        strcpy(path, src_vc);
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c]: That seems needlessly complicated. Why not just use `argv[1]`? I don't think anything is gained by copying it.

In `main`, undo the change to `vc`, and use something like this:
```C
if (argv[1]) {
  vc = argv[1];
  fd = test_source_fd(vc);
} else {
  fd = find_source_vc(&_vc, &idx);
  vc = _vc;
}
```
where `_vc` is a `_cleanup_free_` variable.

File: src/vconsole/vconsole-setup.c [hunk 24]
```diff
@@ -305,36 +300,109 @@ static void setup_remaining_vcs(int fd, bool utf8) {
...
+                fd = -1;
+                return ret_fd;
+        }
+
+        errno = loop_errno;
+        free(path);
+        return log_error_errno(errno, "No usable source console found, last error for /dev/tty%u was: %m", i - 1);
+}
+
+static int test_source_vc(char **ret_path, const char *src_vc) {
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c]: I'd prefer `verify_source_vc`, 'cause `test_` sounds too much like unit testing…

File: src/vconsole/vconsole-setup.c (hunk 19 above)
Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c]: The suggestion was to return fd ....

File: src/vconsole/vconsole-setup.c (hunk 20 above)
Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c]: In this particular function I didn't use cleanup attribute as that feel an overkill for this. Can do both ways of course.

File: src/vconsole/vconsole-setup.c (hunk 22 above)
Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c]: I changed is_settable() to follow coding style more closely. < 0 on error (which in case of == 0 is now EBUSY).

File: src/vconsole/vconsole-setup.c (hunk 19 above)
Comment (keszybz) [reply src/vconsole/vconsole-setup.c]: You're right, it's OK. I though the fd is also returned in one of the output parameters.

File: src/vconsole/vconsole-setup.c (hunk 15 above)
Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c]: ok

File: src/vconsole/vconsole-setup.c (hunk 14 above)
Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c]: git grep ' errno = ' shows near 200 cases, so I assumed it was fine - especially if the caller has to act in the usual fashion (so it would expect errno to be sensibly set when the result is < 0 and errno may be passed down from some other function).

I changed is_settable() (and other functions that previously returned true/false) to follow < 0 / >= 0. Indirectly following Lennarts suggestions.

File: src/vconsole/vconsole-setup.c (hunk 20 above)
Comment (keszybz) [reply src/vconsole/vconsole-setup.c]: Yeah, we prefer to use if for consistency, unless the manual version is shorter or simpler.

File: src/vconsole/vconsole-setup.c (hunk 23 above)
Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c]: In the first review Lennart said it was ok to allocate memory (originally I returned pointer to static array in the function searching for console - or - if argv[1] was present, didn't change the pointer at all; IMHO that was the most simple approach). With the above we get 2 variables serving near same purpose (and still don't avoid allocation).

File: src/vconsole/vconsole-setup.c (hunk 22 above)
Comment (keszybz) [reply src/vconsole/vconsole-setup.c]: Right, you even wrote that in the commit message. In the commit message it sounded OK, but here not so much. Having an `is_` function that can never return false it inconsistent with the rest of the code and doesn't feel particularly clear.

File: src/vconsole/vconsole-setup.c (hunk 23 above)
Comment (keszybz) [reply src/vconsole/vconsole-setup.c]: OK, if you want to allocate it, that's fine too. But just use `strdup`.

File: src/vconsole/vconsole-setup.c (hunk 14 above)
Comment (keszybz) [reply src/vconsole/vconsole-setup.c]: errno is a dismal interface. We use errno when external interfaces give us on other choice (glibc is particularly bad in this regard), but when we some error code in errno, we convert it to local negative error value, and never use errno to pass state internally.

File: src/vconsole/vconsole-setup.c (hunk 22 above)
Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c]: So better to revert those functions to how they were previously - just false / true as their return values ?

File: src/vconsole/vconsole-setup.c [hunk 25]
```diff
@@ -82,7 +84,10 @@ static bool is_settable(int fd) {
          *
          * http://lists.freedesktop.org/archives/systemd-devel/2013-February/008573.html
          */
-        return r == 0 && IN_SET(curr_mode, K_XLATE, K_UNICODE);
+        if (r < 0)
+                return -errno;
+
+        return IN_SET(curr_mode, K_XLATE, K_UNICODE) ? 0 : -EBUSY;
```
Comment (keszybz) [inline src/vconsole/vconsole-setup.c:90]: So this can never return true. OK for a `verify_…` function.

**Comment by keszybz** [comment]
Thanks.
